lib/less/functions/list.js

```diff
@@ -0,0 +1,82 @@
+var Dimension = require('../tree/dimension'),
+    Declaration = require('../tree/declaration'),
+    Ruleset = require('../tree/ruleset'),
+    Selector = require('../tree/selector'),
+    Element = require('../tree/element'),
+    functionRegistry = require('./function-registry');
+
+var getItemsFromNode = function(node) {
+    // handle non-array values as an array of length 1
+    // return 'undefined' if index is invalid
+    var items = Array.isArray(node.value) ?
+        node.value : Array(node);
+
+    return items;
+};
+
+functionRegistry.addMultiple({
+    _SELF: function(n) {
+        return n;
+    },
+    extract: function(values, index) {
+        index = index.value - 1; // (1-based index)
+
+        return getItemsFromNode(values)[index];
+    },
+    length: function(values) {
+        return new Dimension(getItemsFromNode(values).length);
+    },
+    each: function(list, ruleset) {
+        var i = 0, rules = [], rs, newRules, iterator;
+
+        rs = ruleset.ruleset;
+        if (list.value) {
+            if (Array.isArray(list.value)) {
+                iterator = list.value;
+            } else {
+                iterator = [list.value];
+            }
+        } else if (list.ruleset) {
+            iterator = list.ruleset.rules;
+        } else if (Array.isArray(list)) {
+            iterator = list;
+        } else {
+            iterator = [list];
+        }
+
+        iterator.forEach(function(item) {
+            i = i + 1;
+            newRules = rs.rules.slice(0);
+            var key, value;
+            if (item instanceof Declaration) {
+                key = typeof item.name === 'string' ? item.name : item.name[0].value;
+                value = item.value;
+            } else {
+                key = new Dimension(i);
+                value = item;
+            }
+            newRules.push(new Declaration('@value',
```

Comment:
Since these are special reserved variables now, I wonder if we make them less likely to conflict with variable names developers might be using?  Like `@__value` or `@VALUE`?  Is there precedent for reserved variable names?

On a related note, I know "key" and "index" have subtle flavor differences, but I wonder if we minimize reserved variables by just using either one or the other.  Is there a case where we'd need both that I'm overlooking?

Reply:
Well: 

a) These vars are only bound specifically within the context of this function, so how would this possibly conflict?
b) Less already binds an @arguments in the context of mixins, so it would be strange to start suddenly namespacing injected vars in that way.

> On a related note, I know "key" and "index" have subtle flavor differences, but I wonder if we minimize reserved variables by just using either one or the other. Is there a case where we'd need both that I'm overlooking?

You can iterate an `each()` over a ruleset, which will then have, for each entry (declaration), a `key`, `value`, and `index` (numerical position). Unless it should be `@name` instead of `@key` ?

Reply:
> Less already binds an `@arguments` in the context of mixins...

Precedent wins for me.

I think `@name` is probably more likely to have some global value assigned to it, so I think `@key` is better.

TL;DR: I'm convinced.  No changes needed here.

Reply:
Along similar lines, I think better to be explicit (`@index` vs `@i`, and also precedent of full names)

Reply:
Besides; if an author desperately wants a short-hand because their loop-body contains long lines of code they want to condense: nothing is preventing them from doing a quick `@i : @index`...

Reply:
> And for formal parameters, it's better to have formal names.

Yep.